### PR TITLE
feat(nodes): add prompt template node with variable substitution

### PR DESCRIPTION
**Pull request by charliegillet**
## Summary
- Add a new `prompt_template` node that renders prompt templates with dynamic variable substitution from pipeline context
- Supports `{{variable}}` placeholders, built-in helpers (`{{now}}`, `{{uuid}}`, `{{random}}`), conditional sections (`{% if %}`/`{% elif %}`/`{% else %}`/`{% endif %}`), and loop support (`{% for item in list %}`)
- Lightweight template engine implemented with stdlib only — zero external dependencies

## Implementation
| File | Purpose |
|------|---------|
| `services.json` | Node config — lanes (text, questions), fields (template, variables), profiles |
| `IGlobal.py` | Global init — loads config via `Config.getNodeConfig` |
| `IInstance.py` | Instance logic — collects text/questions, renders template on closing |
| `template_engine.py` | Standalone template engine with tokenizer, block eval (if/for), variable resolution |
| `requirements.txt` | Empty — stdlib only |

## Test plan
- [ ] Verify `{{variable}}` substitution renders correctly from configured variables
- [ ] Verify built-in helpers (`{{now}}`, `{{uuid}}`, `{{random}}`) produce valid output
- [ ] Verify `{% if %}` / `{% else %}` / `{% endif %}` conditional rendering
- [ ] Verify `{% for item in list %}` loop rendering
- [ ] Verify escaped braces `\{\{` render as literal `{{`
- [ ] Verify node registers as filter with text and questions lanes
- [ ] Run `ruff format` and `ruff check` — all passing

#Hack-with-bay-2

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a Prompt Template node to render dynamic content with variable interpolation, conditionals (if/elif/else) and loops.
  * Built-in helpers for timestamps, IDs and random values.
  * Handles both plain text and question streams, using configurable templates and variables; default template is {{input}}.

* **Tests**
  * Comprehensive test suite validating rendering behavior, edge cases and safety/resilience.

### CODE REVIEW

**Review by asclearuc — CHANGES_REQUESTED** [review]
Thank you for this contribution — the template engine is well thought out. The tokenizer, recursive block evaluator, and nested depth tracking are all solid, and avoiding a Jinja2 dependency is the right call for this scope.

A few things to address before merging:

**Remove `"gpu"` from `capabilities`**

This node does pure string rendering — no ML, no tensors, no GPU operations. Declaring `"gpu"` will cause the pipeline scheduler to require a GPU-equipped machine for this node, wasting resources. Remove it.

**`_get_template_context` mutates the shared config dict**

`variables` is a static dict set once in node configuration and stored in `self.IGlobal.config`, shared across all instances for the entire pipeline run. `_get_template_context` returns a direct reference to it rather than a copy:

```python
variables = config.get('variables', {})   # live reference, not a copy
variables.setdefault('input', '...')      # mutates shared config
return variables                          # caller mutates it further

Please pay attention also to CodeRabbit's comments

**Comment by charliegillet** [comment]
Addressed all review feedback:
- Removed `"gpu"` from capabilities (addressed in prior commit)
- Fixed `_get_template_context` config dict mutation by copying with `dict()` (addressed in prior commit)
- Added guard against `None` config in `_get_template_context`, `writeQuestions`, and `closing` to prevent `AttributeError` during CONFIG mode (CodeRabbit finding)

Thanks for the thorough review!

File: nodes/src/nodes/prompt_template/IInstance.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        pass
+
+    def _get_template_context(self) -> dict:
+        """Build the context dictionary for template rendering."""
+        config = self.IGlobal.config
+        if config is None:
+            config = {}
+        variables = dict(config.get('variables', {}))
+        if not isinstance(variables, dict):
+            variables = {}
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/IInstance.py]: Fixed — now validates the raw value with isinstance before calling dict(), preventing TypeError when variables is None.

File: nodes/src/nodes/prompt_template/IInstance.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    def writeQuestions(self, question: Question):
+        config = self.IGlobal.config or {}
+        template = config.get('template', '{{input}}')
+        context = self._get_template_context()
+
+        for q in question.questions:
+            context['question'] = q.text
+            rendered = render(template, context)
+            self.question.addQuestion(rendered)
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/IInstance.py]: Fixed — question rendering is now deferred to closing() so all text from writeText() is available in the template context before questions are rendered.

File: nodes/src/nodes/prompt_template/template_engine.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    parts = name.split('.')
+    value = context
+    for part in parts:
+        if isinstance(value, dict):
+            value = value.get(part, _MISSING)
+        else:
+            value = getattr(value, part, _MISSING)
+        if value is _MISSING:
+            return ''
+    return value
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/template_engine.py]: No change needed — the current code already checks context first (lines 72-78) and only falls back to builtins when the name is not found in context (line 80). Users can override {{now}}, {{uuid}}, etc. with their own context values.

File: nodes/src/nodes/prompt_template/IInstance.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            # If we have questions, output them
+            if self.question.questions:
+                self.instance.writeQuestions(self.question)
+            # If we only have text input, render and output as text
+            elif self.collected_text:
+                rendered = render(template, context)
+                self.instance.writeText(rendered)
+
+        except Exception as e:
+            debug(f'Error in prompt_template node: {e}')
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/IInstance.py:84]: The except block now re-raises after logging (see line 87). This was addressed in an earlier commit.

File: nodes/src/nodes/prompt_template/IInstance.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        config = self.IGlobal.config
+        variables = dict(config.get('variables', {}))
+        if not isinstance(variables, dict):
+            variables = {}
+
+        # Add collected text as 'input' if available
+        if self.collected_text:
+            variables.setdefault('input', '\n'.join(self.collected_text))
+
+        return variables
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/IInstance.py:55]: Already addressed — `_get_template_context` and `closing()` both guard against None config with `config = self.IGlobal.config or {}`.

**Comment by charliegillet** [comment]
@asclearuc Addressed all review feedback:

**Fixed in this push:**
- Removed unrelated `PageStatus.tsx` and `styles.css` changes
- Fixed `_get_template_context` to validate `variables` type before copying, preventing TypeError when value is None
- Deferred question rendering to `closing()` so templates have the full text context (all `writeText()` calls) before rendering

**Previously addressed:**
- Removed `"gpu"` from capabilities (already done in earlier commit)
- Config dict mutation: `dict()` copy already ensures shared config is not mutated
- Exception re-raise after logging already in place

**CodeRabbit comments reviewed:**
- Builtins precedence: no change needed — context is checked first, builtins are fallback only
- All other suggestions addressed or already handled

**Comment by nihalnihalani** [comment]
## Senior Review — PR #573: feat(nodes): add prompt template node

Nice work here, @charliegillet. You've been thorough in addressing asclearuc's feedback across both rounds, and the CodeRabbit items have been handled as well. The implementation looks solid.

**Status:** All previously requested changes appear to be addressed. This looks ready for a re-review to verify and move toward approval.

**One item to double-check:** In the `closing()` method, please confirm that exceptions are re-raised after logging. If the current pattern is:

```python
except Exception as e:
    logger.error(...)
```

it should be:

```python
except Exception as e:
    logger.error(...)
    raise
```

Swallowing exceptions during cleanup can mask resource leaks or downstream failures. If this is already handled, disregard.

Overall, this is in great shape. Requesting a re-review from the maintainers to get this across the finish line.

**Comment by nihalnihalani** [comment]
## 🚀 Merge Request

All feedback from asclearuc has been addressed across 2 rounds. Author has been thorough and responsive.

**Status:** All review items resolved. Ready for re-review and merge. @asclearuc

File: nodes/src/nodes/prompt_template/IInstance.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/IInstance.py:84]: Confirmed — this was fixed in commit 6c5f7f5. The exception is now re-raised after logging via `debug()`, so errors propagate correctly to the pipeline engine.

File: nodes/src/nodes/prompt_template/services.json [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+		"text"
+	],
+	//
+	// Required:
+	//		Capabilities are flags that change the behavior of the underlying
+	//		engine
+	//
+	"capabilities": [
+		"gpu"
+	],
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/services.json]: Fixed — `"gpu"` was removed from `capabilities` in commit 6c5f7f5. The array is now empty (`[]`) since this node only does string rendering and has no GPU requirements.

File: nodes/src/nodes/prompt_template/services.json [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+					"questions": [{"text": "Hello world"}]
+				},
+				"expect": {
+					"questions": {
+						"notEmpty": true
+					}
+				}
+			}
+		]
+	},
```
Comment (charliegillet) [inline nodes/src/nodes/prompt_template/services.json:161]: Addressed in c5b406c — added a second test case that validates actual template substitution. The new case configures a template with `{{input}}` variable, provides a concrete input value, and asserts the rendered output contains the expected prefix text.

File: nodes/src/nodes/prompt_template/template_engine.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/template_engine.py]: Good catch — fixed in c5b406c. Restructured `_resolve()` to clearly separate the two resolution phases: (1) walk the context dict first, (2) fall back to builtins only when the context lookup fails. The code now has explicit comments documenting this priority order so the intent is unambiguous.

File: nodes/src/nodes/prompt_template/IInstance.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/IInstance.py:55]: Confirmed — this was fixed in commit b7a26fb. `config` is now guarded with `or {}` before accessing keys, and `_get_template_context` also handles `config is None` explicitly.

File: nodes/src/nodes/prompt_template/IInstance.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/IInstance.py]: Confirmed — fixed in commit a0c39bb. `raw_variables` is now copied via `dict(raw_variables)` before any mutation, so the shared config dict is never modified. The `isinstance` check also guards against non-dict values.

File: nodes/src/nodes/prompt_template/IInstance.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/IInstance.py]: Confirmed — this was addressed in commit a0c39bb. Questions are now rendered in `closing()` after all text has been collected, ensuring the full context (including `input`) is available for template rendering.

**Comment by charliegillet** [comment]
@asclearuc Thanks for the review! All feedback has been addressed:

1. **`"gpu"` removed from `capabilities`** — Already fixed in 6c5f7f5. The array is now empty since this node does pure string rendering.
2. **`_get_template_context` no longer mutates shared config** — Fixed in a0c39bb. `raw_variables` is shallow-copied via `dict()` before any modification, and an `isinstance` guard protects against non-dict values.
3. **Builtins precedence fixed** — Fixed in c5b406c. `_resolve()` now clearly tries context first, then falls back to builtins only when the context lookup fails. The two phases are separated with comments.
4. **More comprehensive test case added** — c5b406c adds a test that configures a real template (`Summarize the following: {{input}}`), provides concrete variable values, and asserts the rendered output contains the expected text.

**Comment by asclearuc** [comment]
Thanks for the work here.

One blocker before this can merge: there are no tests for `template_engine.py` (sorry, noticed it just today).
Given this is a custom template engine (stdlib only, no Jinja2), test coverage is
essential — bugs here will silently corrupt prompts in production pipelines.

At minimum, the tests should cover all the functions from `nodes/src/nodes/prompt_template/template_engine.py` .

File: nodes/src/nodes/prompt_template/services.json [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+	//
+	"classType": [
+		"text"
+	],
+	//
+	// Required:
+	//		Capabilities are flags that change the behavior of the underlying
+	//		engine
+	//
+	"capabilities": [],
```
Comment (asclearuc) [inline nodes/src/nodes/prompt_template/services.json]: Let mark this node as an experimental one.
```suggestion
	"capabilities": ["experimental"],
```

**Comment by charliegillet** [comment]
@asclearuc Added comprehensive unit tests for `template_engine.py` in `nodes/test/test_prompt_template_engine.py` — 57 tests covering variable substitution (flat + dotted paths), missing variables, built-in helpers (`now`/`uuid`/`random`), context overriding builtins (regression guard), conditionals (if/elif/else, nested, empty/missing conditions), for loops (flat + nested, dict iteration, missing iterables), escape sequences, unclosed / malformed tags, HTML pass-through, and ReDoS / recursion safety. All 57 tests pass in ~0.1s. The tests load `template_engine.py` directly via `importlib` so they avoid the `depends`-based package init and can run anywhere Python is available.

Also addressed the remaining feedback in the same commit (6538ad4):
- Marked the node as `"experimental"` in `services.json` per your inline suggestion
- Added an explicit `{{question}}` substitution test case in `services.json` (CodeRabbit)
- Added missing return type annotations on `_resolve`, `_at_top_level`, `_handle_if`, `_handle_for` to satisfy Ruff ANN202 (CodeRabbit)

File: nodes/src/nodes/prompt_template/services.json (hunk 8 above)
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/services.json]: Fixed — marked the node as `"experimental"` in services.json. Commit 6538ad4.

File: nodes/src/nodes/prompt_template/services.json [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+				"questions": {
+					"questions": [{"text": "test prompt"}]
+				},
+				"expect": {
+					"questions": {
+						"notEmpty": true,
+						"contains": "Summarize the following:"
+					}
+				}
+			}
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/services.json:159]: Fixed — added a third test case in services.json that asserts `{{question}}` substitution explicitly: template `Q={{question}} | I={{input}}`, question `what is this?`, expects output to contain `Q=what is this?`. This exercises the per-question substitution path in `IInstance.closing()`. Commit 6538ad4.

File: nodes/src/nodes/prompt_template/template_engine.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+    'now': lambda: datetime.now(timezone.utc).isoformat(),
+    'uuid': lambda: str(_uuid.uuid4()),
+    'random': lambda: str(_random()),
+}
+
+
+# ---------------------------------------------------------------------------
+# Context value resolution
+# ---------------------------------------------------------------------------
+def _resolve(name: str, context: dict):
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/template_engine.py]: Fixed — added return type annotations: `_resolve(...) -> Any`, `_at_top_level(...) -> bool`, `_handle_if(...) -> int`, `_handle_for(...) -> int`, and imported `Any` from `typing`. `ruff check` now passes cleanly on the prompt_template package. Commit 6538ad4.

File: nodes/src/nodes/prompt_template/template_engine.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/prompt_template/template_engine.py]: Thanks for verifying the builtin/context fix! On the Ruff ANN202 follow-up — fixed in commit 6538ad4. Added return type annotations to `_resolve(...) -> Any`, `_at_top_level(...) -> bool`, `_handle_if(...) -> int`, and `_handle_for(...) -> int` (imported `Any` from `typing`). I also added a full unit test suite at `nodes/test/test_prompt_template_engine.py` (57 tests) — including explicit regression tests for the context-overrides-builtins case (`TestContextOverridesBuiltins`).

File: nodes/test/prompt_template/test_engine.py [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+        # blow up (ReDoS guard). Timing check is intentionally loose.
+        import time
+
+        template = 'x' * 50_000 + '{{v}}' + 'y' * 50_000
+        start = time.perf_counter()
+        out = render(template, {'v': 'MID'})
+        elapsed = time.perf_counter() - start
+        assert 'MID' in out
+        assert elapsed < 2.0  # generous upper bound
+
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/test/prompt_template/test_engine.py:348]: Fixed in commit 28a9e9ed — removed the wall-clock `elapsed < 2.0` assertion (and any other timing-based checks) since it could flake on slow/busy runners. The test now just verifies that the tokenizer regex handles the large literal input correctly by asserting the exact expected output. A catastrophic regex backtrack would manifest as a multi-minute hang, which pytest would terminate anyway — so the timing-based check was redundant for the ReDoS guard.

**Comment by charliegillet** [comment]
## Latest fixes

Commit `28a9e9e`:

- Removed the wall-clock `elapsed < 2.0` assertion from `test_large_literal_input_runs_quickly` (per CodeRabbit). The test now asserts the exact expected output of the 100KB render, which still catches a ReDoS regression (catastrophic backtracking would hang for minutes and be killed by pytest) while eliminating CI flakiness on slow/busy runners. No other wall-clock thresholds exist in the test file.

Commit `6538ad4`:

- Added comprehensive unit tests for `template_engine.py` at `nodes/test/test_prompt_template_engine.py` — **57 tests**, all passing in ~0.1s. Covers:
  - Variable substitution (flat + dotted paths + attribute access)
  - Missing variables (graceful empty-string fallback)
  - Built-in helpers: `{{now}}`, `{{uuid}}`, `{{random}}`
  - Context overriding builtins (regression guard for the recent fix)
  - Conditionals: `if` / `elif` / `else`, nested, empty/missing conditions, dotted conditions
  - Loops: `for`, nested `for`, iteration over dicts, missing iterables, string-iterable guard, loop-var shadowing
  - Escape sequences (`\\{{` / `\\}}`)
  - Unclosed / malformed tags (graceful degradation)
  - HTML pass-through (no implicit escaping)
  - Safety: ~100KB input (ReDoS guard), 500-placeholder template, 50-deep `if` nesting, 10-deep `for` nesting
- Marked the node as `"experimental"` in `services.json` (per asclearuc)
- Added a third `services.json` test case that asserts `{{question}}` substitution explicitly (per CodeRabbit)
- Added missing return type annotations on `_resolve` (`Any`), `_at_top_level` (`bool`), `_handle_if` (`int`), `_handle_for` (`int`) to satisfy Ruff ANN202 (per CodeRabbit)
- `ruff check` and `ruff format --check` both pass cleanly on the prompt_template package and the new test file

**Comment by charliegillet** [comment]
@asclearuc Thanks for the thorough reviews. Everything you raised is addressed on the current HEAD (28a9e9e):

1. `"gpu"` capability — removed in 6c5f7f5, then replaced with `["experimental"]` per your inline suggestion in 6538ad4.
2. `_get_template_context` no longer mutates the shared config dict — in a0c39bb, `raw_variables` is copied via `dict(raw_variables)` before any `setdefault`/mutation. An `isinstance(raw_variables, dict)` guard also protects against `None` / non-mapping values.
3. Custom template engine test coverage — added in 6538ad4 at `nodes/test/test_prompt_template_engine.py`. 57 tests, all passing in ~0.04s on Python 3.12. Coverage:
   - Variable substitution (flat + dotted paths + attribute access)
   - Missing variables (graceful empty-string fallback)
   - Built-in helpers (`{{now}}`, `{{uuid}}`, `{{random}}`) and regression guard that context overrides them
   - Conditionals (`if` / `elif` / `else`, nested, empty/missing conditions, dotted conditions)
   - Loops (`for`, nested `for`, iteration over dicts, missing iterables, string-iterable guard, loop-var shadowing)
   - Escape sequences (`\\{{` / `\\}}`)
   - Unclosed / malformed tags (graceful degradation)
   - HTML pass-through (no implicit escaping)
   - Safety: ~100KB input (ReDoS guard), 500-placeholder template, 50-deep `if` nesting, 10-deep `for` nesting

The tests load `template_engine.py` directly via `importlib` so they avoid the `depends`-based package init and can run anywhere Python is available — no server build required.

Ready for re-review whenever you have a moment.

**Comment by charliegillet** [comment]
@nihalnihalani Thanks for the senior review. To confirm your one double-check: yes, `closing()` re-raises after logging. The current pattern at `nodes/src/nodes/prompt_template/IInstance.py:83-85` is:

```python
except Exception as e:
    debug(f'Error in prompt_template node: {e}')
    raise
```

Fixed in 6c5f7f5 per CodeRabbit's earlier finding. Exceptions propagate to the pipeline engine as expected.

**Review by asclearuc — DISMISSED** [review]

**Comment by stepmikhaylov** [comment]
The branch rebased to force ci/cd workflow to run all nodes unit tests.

**Comment by stepmikhaylov** [comment]
CI/CD checks failed

**Review by stepmikhaylov — APPROVED** [review]

**Comment by Rod-Christensen** [comment]
Hi @charliegillet — great work on this! The template engine is well-designed — clean separation, stdlib-only, solid test coverage with safety/ReDoS guards. We've rebased the branch onto current `develop` (clean, no conflicts) and opened a new PR from the rebased branch so CI can run against the latest codebase. Your authorship is preserved in the commits.

Closing this in favor of the new PR (linked below). Thanks again!
